File: data/avatar/gpt commented/Python/codeforces_79_A.py
# Read two integers from input and assign them to x and y
x, y = list(map(int, input().split()))

# Initialize a variable 'a' to keep track of whose turn it is (0 or 1)
a = 0

# Loop while the conditions for valid moves are met
while x >= 0 and y >= 22 or x >= 1 and y >= 12 or x >= 2 and y >= 2:
    # Check whose turn it is based on the value of 'a'
    if a:
        # If it's player 1's turn (a == 1)
        if y >= 22:
            # Player 1 can subtract 22 from y
            y -= 22
        elif x >= 1 and y >= 12:
            # Player 1 can subtract 1 from x and 12 from y
            x -= 1
            y -= 12
        else:
            # Player 1 can subtract 2 from x and 2 from y
            x -= 2
            y -= 2
    else:
        # If it's player 0's turn (a == 0)
        if x >= 2 and y >= 2:
            # Player 0 can subtract 2 from x and 2 from y
            x -= 2
            y -= 2
        elif x >= 1 and y >= 12:
            # Player 0 can subtract 1 from x and 12 from y
            x -= 1
            y -= 12
        else:
            # Player 0 can subtract 22 from y
            y -= 22
            
    # Switch turns between player 0 and player 1
    a = 1 - a

# Print the winner based on the last value of 'a'
# If 'a' is 0, player 1 (Ciel) wins; if 'a' is 1, player 0 (Hanako) wins
print(['Ciel', 'Hanako'][1 - a])

File: data/avatar/gpt commented/Java/codeforces_79_A.java
import java.io.BufferedReader; 
import java.io.IOException; 
import java.io.InputStreamReader; 
import java.util.StringTokenizer; 

public class Main { 

    // Method to check if the required resources can be taken
    public static boolean canTake(int xNeeded, int xAvailable, int yNeeded, int yAvailable) { 
        // Check if the needed x resources are greater than available
        if (xNeeded > xAvailable) return false; 
        // Check if the needed y resources are greater than available
        if (yNeeded > yAvailable) return false; 
        // If both checks pass, return true
        return true; 
    } 

    public static void main(String[] args) throws IOException { 
        // Initialize BufferedReader to read input
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 
        // Tokenize the input line to extract x and y values
        StringTokenizer st = new StringTokenizer(br.readLine()); 
        int x = Integer.parseInt(st.nextToken()); // Parse x value
        int y = Integer.parseInt(st.nextToken()); // Parse y value
        int turn = 0; // Initialize turn counter

        // Infinite loop to alternate turns between two players
        while (true) { 
            // Check if it's the first player's turn (even turn number)
            if (turn % 2 == 0) { 
                // Try to take resources according to the defined rules
                if (canTake(2, x, 2, y)) { 
                    x -= 2; // Deduct resources for the first option
                    y -= 2; 
                } else if (canTake(1, x, 12, y)) { 
                    x -= 1; // Deduct resources for the second option
                    y -= 12; 
                } else if (canTake(0, x, 22, y)) { 
                    y -= 22; // Deduct resources for the third option
                } else { 
                    // If no options are available, first player loses
                    System.out.println("Hanako"); 
                    return; 
                } 
            } else { 
                // Check if it's the second player's turn (odd turn number)
                if (canTake(0, x, 22, y)) { 
                    y -= 22; // Deduct resources for the first option
                } else if (canTake(1, x, 12, y)) { 
                    x -= 1; // Deduct resources for the second option
                    y -= 12; 
                } else if (canTake(2, x, 2, y)) { 
                    x -= 2; // Deduct resources for the third option
                    y -= 2; 
                } else { 
                    // If no options are available, second player loses
                    System.out.println("Ciel"); 
                    return; 
                } 
            } 
            turn++; // Increment turn counter for the next iteration
        } 
    } 
}
